src/meta_memcache/base/connection_pool.py

```diff
@@ -116,28 +130,37 @@ def _discard_connection(self, conn: MemcacheSocket, error: bool = False) -> None
     @contextmanager
     def get_connection(self) -> Generator[MemcacheSocket, None, None]:
         try:
-            conn = self._pool.popleft()
+            conn_holder = self._pool.popleft()
         except IndexError:
-            conn = None
+            conn_holder = None
 
-        if conn is None:
-            conn = self._create_connection()
+        if conn_holder is None:
+            time.sleep(self._new_connection_spacing_s)            
+            conn_holder = self._create_connection()
 
+        should_discard_from_error = False
         try:
-            yield conn
+            yield conn_holder.connection
         except Exception as e:
-            # Errors, assume connection is in bad state
-            _log.warning(
-                "Error during cache conn context (discarding connection)",
-                exc_info=True,
-            )
-            self._discard_connection(conn, error=True)
+            conn_holder.connection_error_count += 1
+
+            should_discard_from_error = conn_holder.connection_error_count >= self._errors_required_to_discard_connection
```

Comment:
The problem is that we can no longer trust the state of the socket when we get an error.

If we read and get a timeout, you can't send anything else on the same socket. The next command might get the response of the first request that timed out the first time.